### DISCUSSION TITLE
Automatic unit conversion for certain sensor device classes

### DISCUSSION THREAD

**Discussion by epenet** [home-assistant/architecture#809] [Entity Models] [answered]
**Summary**
Home assistant should automatically convert distances, pressures, speeds, volumes and weights in a similar way as is done for temperatures.

**Background**
Currently, temperatures in °C will be converted to °F automatically on all sensors which have device class `temperature` and where the unit system is imperial. The opposite is also valid when the unit system is metric. Users do not need to amend manually the unit for all their temperature sensors, and developers do not need to add conversion to their component.

However, this doesn't currently apply to distances, pressures, speeds and volumes, where some components do manual conversions based on `self.hass.config.units`. For components which do not offer automatic conversion, the new device classes allow for user override, but it would be tedious if you have lots of sensors.

Taking distance as an example, it would be good if we had some way to **automatically** convert `cm <-> in`, `m <-> ft` and `km <-> mi` (though we would still need to take into account `mm` and `yd`)

Linked discussions:
- support for UK volume units: #813
- custom unit systems: #814

**Comment by frenck** [comment]
I don't understand, could you extend the proposal?

**Reply by epenet** [reply]
I have updated the proposal.

**Comment by frenck** [comment]
For temperature I think it makes sense, but for things like distance and speed, that is really hard / problematic.

For example, in various countries, the default units/system changes based on what it is. For example, the difference between wind speed vs. car speed. Those can be in completely different systems and units.

**Reply by epenet** [reply]
Agreed, and that is what this discussion is about. Are there units where we think it is useful?
For example, should `miles` always be converted to `kilometers`, and `feet` to `meters` on a metric system? Should `kilometers` and `meters` always be converted to `miles` and `feet` on an imperial system?

Should we allow users to define a custom list of "automatic conversion" listing all convertible units, and maybe deprecate the UnitSystem completely?

For example, for distances I might want to customize with:
- `mm → None` don't convert automatically, but let me choose per-sensor if needed
- `m → None` don't convert automatically, but let me choose per-sensor if needed
- `km → None`: don't convert automatically, but let me choose per-sensor if needed

- `cm → mm`  automatic conversion because I don't like cm, but let me override per-sensor

- `in → cm`: automatic conversion because I don't know imperial units, but let me override per-sensor
- `ft → m`: automatic conversion because I don't know imperial units, but let me override per-sensor
- `yd → m`: automatic conversion because I don't know imperial units, but let me override per-sensor
- `mi → km`: automatic conversion because I don't know imperial units, but let me override per-sensor

**Reply by frenck** [reply]
This makes all kinds of assumptions and adds "magic".  That is IMHO a no-go.

**Reply by epenet** [reply]
Quite the opposite, this solution wouldn't make any assumptions, and give 100% control to the users.
Today, users can only choose between Metric and Imperial. Tomorrow then could customize it to their liking.

- A "non-customized system" would have all mappings set to `None`
- On the customization screen we could have buttons to "disable all imperial units" that automatically pre-selects metric units for all imperial units, or an equivalent button to "enable all imperial units" that automatically pre-selects imperial units for all metric units, or a reset button to reset all mappings to None.
- We could even have country-specific "presets" that applies a specific set, whilst still allowing the user to adjust based on their specific needs.

**Reply by frenck** [reply]
> Today, users can only choose between Metric and Imperial. Tomorrow then could customize it to their liking.

They already can, this architecture talks about "automatic unit conversion". So it is talking about making assumptions on conversion.

Right now, this is already problematic for temperatures with 3D printers, which we now, for example, just convert to `F`, while even our American friends use `C` for those.

It assumes it is all the same, it should not. That is the whole point. And many countries use different systems depending on the subject.

If there is an architecture issue I would support, it would be removing automatic conversion of `temperature` based on the unit setting.

I have a pretty strong opinion on this proposal and are against the proposal.

**Reply by epenet** [reply]
If I understand correctly, you are favour in deprecating the Unit System completely, and not replace it with anything.
Maybe this would be OK if there is a UI to "override sensors in bulk"

**Reply by frenck** [reply]
> If I understand correctly, you are favour in deprecating the Unit System completely, and not replace it with anything.

Ideally: Yes.

**Comment by frenck** [comment]
> Taking distance as an example, it would be good if we had some way to automatically convert cm <-> in, m <-> ft and km <-> mi (though we would still need to take into account mm and yd)

Not sure why we keep re-inventing the conversions wheel in Home Assistant, as many Python packages are around for these things. Most notably `pint`.

**Reply by epenet** [reply]
Now that the utils are centralised in https://github.com/home-assistant/core/blob/dev/homeassistant/util/unit_conversion.py, I plan to look at leveraging `pint` to externalise the processes.
Regardless, I don't think that `pint` will help in choosing "the right unit" and makes this discussion still valid.

**Reply by frenck** [reply]
> Regardless, I don't think that pint will help in choosing "the right unit" and makes this discussion still valid.

Sorry, it is valid for this discussion, as it was added to the proposal.

**Reply by epenet** [reply]
Draft PR for `pint` is on https://github.com/home-assistant/core/pull/79290

**Comment by bdr99** [comment]
In my opinion, for any automatic conversion logic that we could attempt to implement, there will always be cases where it is not the unit that the user wanted. See the examples mentioned by @emontnemery in https://github.com/home-assistant/architecture/discussions/682#discussioncomment-3582684.

I think this problem is kind of already solved by the `native_unit_of_measurement` attribute of sensor entities. For sensor entities that have a device class that supports manually changing the unit, such as pressure, the `native_unit_of_measurement` represents the _default_ unit that will be used if the user has not manually changed the unit in the UI. This allows the integration to determine what the best default unit should be, taking into account both the user's HA unit system preference (metric vs. imperial) as well as the specific type of entity in question (such as tire pressure vs. atmospheric pressure). If this is ever not the unit that the user wanted, they can still override it using the unit selection in the UI, since HA still understands that it is a pressure device class.

So I think the best path forward is for integrations to just set the `native_unit_of_measurement` conditionally based on the user's metric/imperial unit system. I do this in the Mazda integration for the odometer entity here: https://github.com/home-assistant/core/blob/dev/homeassistant/components/mazda/sensor.py#L279-L284

This is my understanding about how `native_unit_of_measurement` works, but please correct me if anything I said is incorrect.

**Reply by epenet** [reply]
IMO, `native_unit_of_measurement` should be the `non-converted` raw unit of measurement returned by the API/integration.
Otherwise you could end up with a `kilometers` sensor being automatically converted to `miles` based on the unit system, and then being re-converted back to `kilometers` because the users has customized it on his imperial system.

These are the type of problems that need to be discussed, and why I opened this discussion

**Comment by emontnemery** [comment]


**Reply by frenck** [reply]
Why extract the discussion away from the architectural discussion?

**Comment by emontnemery** [comment]
@balloob suggested that instead of forcing the automatic conversion on sensors, we can let sensors hint which unit they prefer in imperial + metric system. It could perhaps be something like this:

```py
class MyMetricTravelDistanceSensor(SensorEntity):
    """Distance sensor reporting in km/h, hinting mph for users using the imperial system."""
    _attr_device_class = "distance"
    _attr_native_unit_of_measurement = "km/h"
    _attr_imperial_unit_of_measurement = "mph"

    ...

class MyImperialTravelDistanceSensor(SensorEntity):
    """Distance sensor reporting in mph, hinting km/h for users using the metric system."""
    _attr_device_class = "distance"
    _attr_native_unit_of_measurement = "mph"
    _attr_metric_unit_of_measurement = "km/h"

    ...

class MySITravelDistanceSensor(SensorEntity):
    """Distance sensor reporting in m/s, hinting km/h / mph for users using the metric or imperial system."""
    _attr_device_class = "distance"
    _attr_native_unit_of_measurement = "m/s"
    _attr_imperial_unit_of_measurement = "mph"
    _attr_metric_unit_of_measurement = "km/h"

    ...

class My3DPrinterTemperatureSensor(SensorEntity):
    """Temperature sensor reporting in °C, hinting °C also for users using the imperial system.

    Note: Temperatures have always been automatically converted, to avoid a breaking change temperature
    sensors opt out instead of opt in.
    """
    _attr_device_class = "distance"
    _attr_native_unit_of_measurement = "°C"
    _attr_imperial_unit_of_measurement = "°C"

    ...
```

**Reply by frenck** [reply]
I don't see how that works? As depending on country, a mix of imperial and metric is used.
How is this addressing/fixing that?

This still assumes a country is using only metric or only imperial, which is, not the case. I don't see how this is different from the original discussion (besides integrations raising their preference, which can be even a bigger chaos/more inconsistent and thus be wrong potentially).

**Reply by emontnemery** [reply]
> This still assumes a country is using only metric or only imperial, which is, not the case.

Sure, but we don't handle that fact today either, we just allow the user to choose metric or US imperial.
To address that, we should add support for Canadian and UK mixed unit systems, which is a different topic.

> even a bigger chaos/more inconsistent

Can you give some examples of how this proposal would result in bigger chaos and more inconsistency than what we have today?
Do you mean that an integration author potentially picks a unit which is unsuitable for the sensor, let's say km/h for rain fall?

**Reply by thecode** [reply]
I apologize if it was already discussed, I tried to follow all the comments, in my option integrations should not do any unit conversion or suggest a unit, they should only report the native value of the sensor (the actual unit from the device) and the device class. The rest should be handled by core (detailed later), How will it work:

Let's take a wind measurement sensor, The manufacture reports the wind in `m/s` (which is for this example probably not something common) instead of the integration doing conversions it should report the value as it is to core and the applicable unit.
Core will show it to the user based on 3 parameters (in order of importance):
1. If user override the unit the value should be converted to the override unit
2. If a local regional setting apply use that to convert the unit (for example wind speed in `mph`
3. Use the unit correlated to the device class (e.g. `km/h`)

This should give the best user experience, it reduce redundant unit conversations and simplify integrations.
In cases that the device does not provide the data in a valid unit, for example if it is an ADC value and is not handled upstream (which it should) the integration can use any system (preferably metric) and the rest will be handled by core.

As I understand 1 is already implemented, 2 can be done in a later stage and 3 does not require major changes

**Reply by emontnemery** [reply]
@thecode I think that makes sense, adding device classes for common applications (barometric pressure, wind speed, rainfall) would go a long way.

There are still edge cases though; for example American 3D printer users apparently want their 3D printer's temperature sensors to report in °C, while any temperature sensor reporting weather forecast, indoor or outdoor temperature should be converted to °F. Adding a "3D printer temperature sensor" device class for this edge case is probably not reasonable?

**Reply by ryanm101** [reply]
This has appeared as an issue here: https://github.com/home-assistant/core/pull/79223

I think a more 'correct' overall solution would be instead of asking the user imperial or metric add a third option custom. Then show a list of the measurements and let the user specify their preferred default value. Then allowing an override per sensor if required. 

There are some edge cases as well due to age where certain ages will have had different learnings. e.g.
My Dad - Imperial (doesnt understand metric at all)
Me - Was taught metric in school but at home used imperial so while i understand some units of metric i need to use imperial for others (weight is a prime example I have no concept of 1kg but i do understand 1lb. and when ordering sweats would order by the 1/4 lb) would weigh people in st and lb not in lb only like Americans or in kg like Europeans. 
My kids - are being taught metric BUT once they leave school will discover that the UK does not use pure metric, so at the least will have to learn and use miles.

**Reply by thecode** [reply]
> I think a more 'correct' overall solution would be instead of asking the user imperial or metric add a third option custom. Then show a list of the measurements and let the user specify their preferred default value. Then allowing an override per sensor if required.

@ryanm101  I think it is mandatory that things will happen automatically. A user add an integration with an hub that supports multiple devices, it would be a bad user experience if you ask him the units for all the devices supported by that integration when you are not sure they will ever be used.

Your case can be easily covered by letting the user select the regional settings he would like to use (single time at first HA setup) and let him override units for sensors he would like. This is almost implemented, except that today we only allow to choose between metric and imperial but not mixed by regional setting.
![image](https://user-images.githubusercontent.com/1858925/193217106-2cdd0669-35e7-427b-abdd-7390e5471e88.png)

@emontnemery had a discussion about my suggestion above which lakes the option to allow an integration to force a specific unit (e.g. the 3D printer case) so we might need to allow an integration to hint about recommended unit, however that suggestion alone does not solve the case of countries with mixed units.

**Reply by ryanm101** [reply]
@thecode 

When I was talking about custom user preference it wasnt about forcing the user to do it for every sensor. it was about saying here are some classes of measurements e.g. temp, wind speed, distance and having them choose their preferred default between imperial and metric.  THe reason for cusom was to make that 3rd option useful to non-uk mixed settings (not really sure if there are any but I assume there must be one or two) or for people in the generational problem mentioned above. 

The per sensor override is useful for when something like the US 3d printing temp appears above where they would want to change it in a very specific instance. 

@emontnemery I noticed `gal` up there in your list of conversions beaware there are two sizes of `gal` the British one and the American one so local might be needed to default to one or other. Other regional differences might include weight as UK we use st & lb for a persons weight but in the US they would only refer to lb.

**Reply by emontnemery** [reply]
@ryanm101 Home Assistant currently only support US and metric units, there's no support for British imperial units. Hence, gallon is a US gallon, not a British imperial gallon.

**Reply by epenet** [reply]
UK versus US variations on imperial units (for example, `us fl oz = uk fl oz * 0.96076`) should be a separate architecture discussion.
I have opened https://github.com/home-assistant/architecture/discussions/813 to discuss this.

**Reply by thecode** [reply]
@epenet I agree that the question if we should support UK variation is kina another discussion, however they are related and if we support regional settings we can also cover this requirement

I had a chat about all these variation with @emontnemery, so what I think might help to solve the unit conversion is to allow setting unit for entities **in the following steps (first step has higher priority):**
1: user (any setting the user set that overrides the unit)
2: hinting (this needs to be finalized, but something which let the integration recommend a unit for different regional settings)
3: regional (according to device class + unit + unit system), for example convert precipitation intensity in mm/h to in/h if the user has chosen the US imperial system
default: no conversion

We will need to introduce an option to set global regional setting for HA, same as what is implemented in various OS.

**Reply by emontnemery** [reply]
> We will need to introduce an option to set global regional setting for HA, same as what is implemented in various OS.

We already have it, the user can select either the metric or the US systems.

**Reply by thecode** [reply]
> > We will need to introduce an option to set global regional setting for HA, same as what is implemented in various OS.
> 
> We already have it, the user can select either the metric or the US systems.

Yes, but to cover UK variations for example this will need to be changed (on the same place it already exists) from two radio buttons that only allow metric or US to a list that we expand with the time for different reginal settings. I also noticed now we have another place for setting number format which would be probably better to merge this with the regional settings and would also be better to do it per user and not per system as we do now for units.

**Reply by ryanm101** [reply]
> @ryanm101 Home Assistant currently only support US and metric units, there's no support for British imperial units. Hence, gallon is a US gallon, not a British imperial gallon.

True, though imo it would be better from the get go to list gal as usgal/us_gal or something, else it will potentially cause confusion later if other units are implemented. Especially if someone isn't aware of the difference and chooses the wrong one it could have consequences given liquid is involved. e.g. Beer making putting too little (us gal) in could cause something to burn dry or calculating MPG for a car the US gal being 20% smaller would give a UK car worse MPG than its real value. 
US Gal = 3.785 liters
UK Gal = 4.546 liters

As a side note i didn't actually notice that this could have an effect as prior to my weather data the External system was sending in all units so i could pick and choose now we are talking about auto conversion my (UK) use case becomes more important as if i choose a label saying 'gal' when talking about my car I would expect it to be the correct regional one assuming i lived in blissful ignorance of american specifics.

**Comment by gjohansson-ST** [comment]
I would suggest the complete opposite than the proposal.
Let the integration set the native uom and then let the user override as they prefer (which is already in place in many occassions).
So to take an example from @frenck above I think the one writing the integration for a 3d printer would set celsius as the native uom and if a user prefers he can then override it to fahrenheit.

**Reply by balloob** [reply]
The problem with this approach is that we end up with the same logic in every integration, checking against the unit system in use. And if we add a new unit system, we have to update X integrations.

**Comment by balloob** [comment]
Here is my latest proposal after discussing this a bit offline.

Today we already store the unit of measurement when an entity is registered in the entity registry.

My proposal is to first change it for sensors that we will no longer update the unit of measurement in the entity registry. Instead, the entity value will be converted to whatever is stored in the entity registry. Because it happens when an entity is registered, unit changes at the source will only apply to new entities. 

If an entity had a wrong unit and it needs updating in the registry, they would need to migrate existing entities during startup.

A 2nd step would be to be smarter with a unit system about which unit an entity is registered with. But that should be a standalone proposal.

**Reply by ryanm101** [reply]
What about this suggestion https://github.com/home-assistant/architecture/discussions/813#discussioncomment-3868538

**Comment by emontnemery** [comment, marked as answer]
Automatic unit conversion has been implemented by allowing the unit system to influence the unit used in the state of certain sensor device classes. The changes are summarized in the dev blog post: https://github.com/home-assistant/developers.home-assistant/pull/1561
